### PR TITLE
implement `_CCCL_HAS_INCLUDE`

### LINKED ISSUE #2785: [FEA]: Implement `_CCCL_HAS_INCLUDE`
### Is this a duplicate?

- [x] I confirmed there appear to be no [duplicate issues](https://github.com/NVIDIA/cccl/issues) for this request and that I agree to the [Code of Conduct](CODE_OF_CONDUCT.md)

### Area

General CCCL

### Is your feature request related to a problem? Please describe.

If not defined CCCL defines `__has_include` on various places.

### Describe the solution you'd like

Provide unified `_CCCL_HAS_INCLUDE` macro implemented as:
```cpp
#ifdef __has_include
#  define _CCCL_HAS_INCLUDE(_X) __has_include(_X)
#else
#  define _CCCL_HAS_INCLUDE(_X) 0
#endif
```

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by davebayer**
## Description

Closes #2785.

This PR implements `_CCCL_HAS_INCLUDE` in `cuda/std/__cccl/preprocessor.h` module. All appearances of `__has_include` were replaced by this macro. CCCL does not define `__has_include` anymore if not predefined.

## Checklist
- [x] I am familiar with the [Contributing Guidelines]().
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: cub/cub/detail/nvtx.cuh [hunk 1]
```diff
@@ -103,9 +104,11 @@ CUB_NAMESPACE_END
 #    define CUB_DETAIL_NVTX_RANGE_SCOPE_IF(condition, name)
 #    define CUB_DETAIL_NVTX_RANGE_SCOPE(name)
 #  endif // NVTX3_CPP_DEFINITIONS_V1_0
-#else // __has_include(<nvtx3/nvToolsExt.h> ) && !defined(CCCL_DISABLE_NVTX) && !defined(NVTX_DISABLE) && _CCCL_STD_VER
+#else // _CCCL_HAS_INCLUDE(<nvtx3/nvToolsExt.h> ) && !defined(CCCL_DISABLE_NVTX) && !defined(NVTX_DISABLE) &&
+      // _CCCL_STD_VER
       // >= 2014
```
Comment (miscco) [inline cub/cub/detail/nvtx.cuh:109]:
```suggestion
      // _CCCL_STD_VER >= 2014
```

File: cub/cub/detail/nvtx.cuh [hunk 2]
```diff
@@ -103,9 +104,11 @@ CUB_NAMESPACE_END
...
 #  define CUB_DETAIL_NVTX_RANGE_SCOPE_IF(condition, name)
 #  define CUB_DETAIL_NVTX_RANGE_SCOPE(name)
-#endif // __has_include(<nvtx3/nvToolsExt.h> ) && !defined(CCCL_DISABLE_NVTX) && !defined(NVTX_DISABLE) && _CCCL_STD_VER
+#endif // _CCCL_HAS_INCLUDE(<nvtx3/nvToolsExt.h> ) && !defined(CCCL_DISABLE_NVTX) && !defined(NVTX_DISABLE) &&
+       // _CCCL_STD_VER
        // >= 2014
```
Comment (miscco) [inline cub/cub/detail/nvtx.cuh:114]:
```suggestion
       // _CCCL_STD_VER >= 2014
```

File: thrust/thrust/type_traits/remove_cvref.h [hunk 3]
```diff
@@ -31,10 +31,8 @@
 #  pragma system_header
 #endif // no system header
 
-#if _CCCL_STD_VER >= 2017
-#  if __has_include(<version>)
-#    include <version>
-#  endif
+#if _CCCL_HAS_INCLUDE(<version>)
+#  include <version>
 #endif
```
Comment (miscco) [inline thrust/thrust/type_traits/remove_cvref.h:36]:
```suggestion
#endif // _CCCL_HAS_INCLUDE(<version>)
```

File: cub/cub/detail/nvtx.cuh [hunk 4]
```diff
@@ -47,14 +47,15 @@
 // * NVTX is not explicitly disabled (via CCCL_DISABLE_NVTX or NVTX_DISABLE)
 // * C++14 is availabl for cuda::std::optional
 // * NVTX3 uses module as an identifier, which trips up NVHPC
-#if __has_include(<nvtx3/nvToolsExt.h> ) && !defined(CCCL_DISABLE_NVTX) && !defined(NVTX_DISABLE) \
-&& _CCCL_STD_VER >= 2014 &&(!defined(_CCCL_COMPILER_NVHPC) || _CCCL_STD_VER <= 2017)
+#if _CCCL_HAS_INCLUDE(<nvtx3 / nvToolsExt.h>) && !defined(CCCL_DISABLE_NVTX) && !defined(NVTX_DISABLE) \
```
Comment (miscco) [inline cub/cub/detail/nvtx.cuh:50]: I am unsure whether that here is a but or whether __has_include still works
```suggestion
#if _CCCL_HAS_INCLUDE(<nvtx3/nvToolsExt.h>) && !defined(CCCL_DISABLE_NVTX) && !defined(NVTX_DISABLE) \
```

**Review by miscco — APPROVED** [review]
Thanks a  lot

Mostly Nitpicks regarding style

**Review by miscco — CHANGES_REQUESTED** [review]
I played with it a bit more and even `WhitespaceSensitiveMacros(...)` wont stop clang-format messing with the include paths.

I am not sure whether there is a simple solution here. We would either need to add all kinds of clang-format suppression

File: cub/cub/detail/nvtx.cuh (hunk 4 above)
Comment (davebayer) [reply cub/cub/detail/nvtx.cuh:50]: Oh, it must have been done by `clang-format`. I will try to fix it.

File: thrust/thrust/detail/config/memory_resource.h [hunk 5]
```diff
@@ -34,10 +34,10 @@
 #define THRUST_MR_DEFAULT_ALIGNMENT alignof(THRUST_NS_QUALIFIER::detail::max_align_t)
 
 #if _CCCL_STD_VER >= 2017
-#  if __has_include(<memory_resource>)
+#  if _CCCL_HAS_INCLUDE(<memory_resource>)
 #    define THRUST_MR_STD_MR_HEADER <memory_resource>
 #    define THRUST_MR_STD_MR_NS     std::pmr
-#  elif __has_include(<experimental/memory_resource>)
+#  elif _CCCL_HAS_INCLUDE(<experimental / memory_resource>)
```
[the remark this answers is not in the document]
Comment (davebayer) [reply thrust/thrust/detail/config/memory_resource.h:40]: Seems it does not. See [link](https://godbolt.org/z/9zKv51oM8).

File: cub/cub/detail/nvtx.cuh (hunk 4 above)
Comment (davebayer) [reply cub/cub/detail/nvtx.cuh]: Fixed.

File: cub/cub/detail/nvtx.cuh (hunk 2 above)
Comment (davebayer) [reply cub/cub/detail/nvtx.cuh]: Done.

File: thrust/thrust/detail/config/memory_resource.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (davebayer) [reply thrust/thrust/detail/config/memory_resource.h]: That's weird, `WhitespaceSensitiveMacros` works fine on my machine after manually removing the spaces.

[the remark this answers is not in the document]
Comment (miscco) [reply thrust/thrust/detail/config/memory_resource.h]: Lets try and add it and see what pre-commit is doing with it

**Comment by miscco** [comment]
pre-commit.ci autofix

File: thrust/thrust/detail/config/memory_resource.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (davebayer) [reply thrust/thrust/detail/config/memory_resource.h]: Seems to be working now

**Comment by miscco** [comment]
Seems like I need to check my local formatter https://github.com/NVIDIA/cccl/pull/2786/commits/bb3c894c3ac5e533862736fc43085c6df3f3f450

File: cub/cub/detail/nvtx.cuh [hunk 6]
```diff
@@ -47,14 +47,16 @@
...
 // * NVTX3 uses module as an identifier, which trips up NVHPC
-#if __has_include(<nvtx3/nvToolsExt.h> ) && !defined(CCCL_DISABLE_NVTX) && !defined(NVTX_DISABLE) \
-&& _CCCL_STD_VER >= 2014 &&(!defined(_CCCL_COMPILER_NVHPC) || _CCCL_STD_VER <= 2017)
+#if _CCCL_HAS_INCLUDE(<nvtx3/nvToolsExt.h>) && !defined(CCCL_DISABLE_NVTX) && !defined(NVTX_DISABLE) \
+                      && _CCCL_STD_VER >= 2014                                                       \
+                      && (!defined(_CCCL_COMPILER_NVHPC) || _CCCL_STD_VER <= 2017)
 // Include our NVTX3 C++ wrapper if not available from the CTK
-#  if __has_include(<nvtx3/nvtx3.hpp>) // TODO(bgruber): replace by a check for the first CTK version shipping the header
+#  if _CCCL_HAS_INCLUDE(<nvtx3/nvtx3.hpp>) // TODO(bgruber): replace by a check for the first CTK version shipping the
+// header
```
Comment (miscco) [inline cub/cub/detail/nvtx.cuh:55]: Lets move the comment
```suggestion
// TODO(bgruber): replace by a check for the first CTK version shipping the header
#  if _CCCL_HAS_INCLUDE(<nvtx3/nvtx3.hpp>) 
```

**Review by miscco — APPROVED** [review]
Awesome, thanks a lot for that cleanup. Loving it

**Comment by davebayer** [comment]
Hopefully it will not break someone's code relying on `__has_include` to be always defined.

**Comment by miscco** [comment]
> Hopefully it will not break someone's code relying on `__has_include` to be always defined.
There are bugs and then there is this. 

Technically a user should not even utter those words

**Comment by miscco** [comment]
pre-commit.ci autofix

**Comment by miscco** [comment]
Thanks a lot for improving that situation 🎉

**Pull request merged**